libflux/src/semantic/flatbuffers/mod.rs

```diff
@@ -0,0 +1,672 @@
+    arg: (String, MonoType, bool, bool),
+) -> flatbuffers::WIPOffset<fb::Argument<'a>> {
+    let name = builder.create_string(&arg.0);
+    let (buf_offset, typ) = build_type(builder, arg.1);
+    fb::Argument::create(
+        builder,
+        &fb::ArgumentArgs {
+            name: Some(name),
+            t_type: typ,
+            t: Some(buf_offset),
+            pipe: arg.2,
+            optional: arg.3,
+        },
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use maplit;
+
+    #[rustfmt::skip]
+    use crate::semantic::flatbuffers::semantic_generated::fbsemantic::{
+        Expression,
+        ExpressionStatement,
+        ExpressionStatementArgs,
+        File,
+        FileArgs,
+        FloatLiteral,
+        FloatLiteralArgs,
+        Operator,
+        Package,
+        PackageArgs,
+        Statement,
+        UnaryExpression,
+        UnaryExpressionArgs,
+        WrappedStatement,
+        WrappedStatementArgs,
+    };
+
+    fn fb_serde(want: PolyType) {
+        // Init flatbuffer builder
+        let mut fb = flatbuffers::FlatBufferBuilder::new();
+        // Encode PolyType
+        let off = build_polytype(&mut fb, want.clone());
+        // Serialize PolyType
+        fb.finish(off, None);
+        let buf = fb.finished_data();
+        // Deserialize and compare
+        assert_eq!(
+            want,
+            PolyType::from(flatbuffers::get_root::<fb::PolyType>(buf))
+        )
+    }
+
+    #[test]
+    fn serde_basic_types() {
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Bool,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Int,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Uint,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Float,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::String,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Duration,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Time,
+        });
+        fb_serde(PolyType {
+            vars: Vec::new(),
+            cons: HashMap::new(),
+            expr: MonoType::Regexp,
+        });
+    }
+    #[test]
+    fn serde_array_type() {
+        fb_serde(PolyType {
+            vars: vec![Tvar(0)],
+            cons: HashMap::new(),
+            expr: MonoType::Arr(Box::new(Array(MonoType::Var(Tvar(0))))),
+        });
+    }
+    #[test]
+    fn serde_function_types() {
+        fb_serde(PolyType {
```

Comment:
You could use simple strings and the type expression grammar to produce the types here, and that would make the test a lot easier to read (and write too).

Reply:
That's a great idea!